src/cysignals/implementation.c

```diff
@@ -526,3 +537,310 @@ static void sigdie(int sig, const char* s)
  * signals.pyx. These require some of the above functions, therefore
  * this include must come at the end of this file. */
 #include "macros.h"
+
+#else /* Windows version */
+
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+#include <sys/types.h>
+#include <Python.h>
+#include <stdlib.h>
+#include "struct_signals.h"
+#include <windows.h>
+#include <float.h>
+
+#ifndef __MINGW32__
+#  define inline __inline
+#endif
+/* The cysigs object (there is a unique copy of this, shared by all
+ * Cython modules using cysignals) */
+static cysigs_t cysigs;
+static int win32ctrlc;
+
+static void do_raise_exception(int sig);
+static void sigdie(int sig, const char* s);
+static void print_backtrace(void);
+static void setup_cysignals_handlers(void);
+static void cysigs_interrupt_handler(int sig);
+static void cysigs_signal_handler(int sig);
+
+/* Do whatever is needed to reset the CPU to a sane state after
+ * handling a signals.  In particular on x86 CPUs, we need to clear
+ * the FPU (this is needed after MMX instructions have been used or
+ * if an interrupt occurs during an FPU computation).
+ * Linux and OS X 10.6 do this as part of their signals implementation,
+ * but Solaris doesn't.  Since this code is called only when handling a
+ * signal (which should be very rare), it's better to play safe and
+ * always execute this instead of special-casing based on the operating
+ * system.
+ * See http://trac.sagemath.org/sage_trac/ticket/12873
+ */
+static inline void reset_CPU(void)
+{
+    _fpreset();
+}
+
+/* Handler for SIGINT, SIGALRM
+ *
+ * On Windows, the way we handle an interrupt immediately is by
+ * mapping it to SIGFPE.  Specifically, we save the signal number in
+ * cysigs.sig_mapped_to_FPE and raise SIGFPE.  The handler for SIGFPE
+ * will then call longjmp.  The reason that it is done this way is
+ * that for most signals in Windows the handler is run in a separate
+ * thread, with its own stack.  This makes it impossible to call
+ * longjmp from a signal handler.  However, longjmp within a signal
+ * handler is supported for exactly one signal, namely SIGFPE.
```

Comment:
Do you have a reference for this?

Reply:
It would also be good to know more precisely when signal handlers are run in a separate thread. Does it depend on the way how the signal was raised (by pressing CTRL-C versus `raise(SIGINT)` for example) or does it actually depend on the signal number (say, `SIGINT` versus `SIGFPE`)?

Reply:
I have no reference for this. This comment is from cypari cysignals fork.

Reply:
Sorry, but I really do not feel comfortable accepting complicated code without knowing why it needs to be so complicated. This is probably the place where cysignals-for-Windows differs the most from the original cysignals, so this requires a good explanation.

Reply:
Yes indeed, that one thing we should clarify.

Reply:
Reference for handler being in their own thread on windows. https://docs.microsoft.com/en-us/windows/console/handlerroutine